### DISCUSSION TITLE
STM32F401 support

### DISCUSSION THREAD

**Discussion by olhodneland** [grblHAL/core#405] [General]
Hi @terjeio  I'm trying to flash a uStepperS32 with the grblHAL firmware. Seems like it's flashing, but after finished it does not show up as a usb device anymore. Can you please explain how this should be done properly? 

- STM32F401 32-bit 84MHz Cortex®-M4 MCU 
- TMC5130 drivers
https://www.ustepper.com/shop/#!/products/ustepper-s32

I'm a Mac user by the way

**Comment by terjeio** [comment]
Seems like it has a 20MHz crystal, a new build configuration has to be added for that. You are compiling with STM32CubeIDE?

**Comment by olhodneland** [comment]
It seems to be 84MHzThe new uStepper S32 with 32-bit 84MHz Cortex®-M4 MCU, USB-C interface and a vast amount of I/O's is here !

Terje Io ***@***.***> wrote:﻿
Seems like it has a 20MHz crystal, a new build configuration has to be added for that. You are compiling with STM32CubeIDE?

**Comment by olhodneland** [comment]
Product_sheet_S32PDF Document · 3,7 MBSent from my iPhoneOn 13 Dec 2023, at 10:42, ***@***.*** wrote:﻿It seems to be 84MHzThe new uStepper S32 with 32-bit 84MHz Cortex®-M4 MCU, USB-C interface and a vast amount of I/O's is here !

Terje Io ***@***.***> wrote:﻿
Seems like it has a 20MHz crystal, a new build configuration has to be added for that. You are compiling with STM32CubeIDE?

**Reply by terjeio** [reply]
> It seems to be 84MHz

That is the the MCU core frequency which is, [among others](https://github.com/grblHAL/STM32F4xx/blob/5f601d9fe16087ce65b62b2f2dcc791dadebeeea/Src/main.c#L53-L345), derived from the crystal.
The USB peripheral requires a 48 MHz clock, if not correct USB will not work. If you are compiling locally (not using the Web Builder) then I can provide parameters for a 20MHz crystal so you can verify before I commit an update.

> Kan kanskje ta resten av kommunikasjonen på norsk?

I prefer using english here. If we need to go to private emails then we can switch.

**Reply by olhodneland** [reply]
Of course, sorry I was answering from my email so did not think it was on git. 

Ok, yes please if you can provide me with a version/parameters that should work I can test it!

Because, there is no reason that grblHAL should not work on the uStepper S32 since it's using a STM32F401?

**Reply by terjeio** [reply]
> Ok, yes please if you can provide me with a version/parameters that should work I can test it!

How are you compiling the firmware?

> Because, there is no reason that grblHAL should not work on the uStepper S32 since it's using a STM32F401?

Looking closer at the board it seems to me that it is intended for driving a single stepper motor with positional feedback?
So it might not be the best choice for grblHAL?
If enough pins are broken out to the pin headers (including the male one) it could still be used but would require a custom pin map?
Or have you selected one already that can be used?

**Reply by olhodneland** [reply]
I can use whatever to compile it. VisualStudios for example. Or any recommendations?

Yes it is intended for that. But it's also intended to combine several of the boards, one on each motor. So the thought is to install grblHAL on one of the boards and use it as a master, and run the other boards (for the other motors) as slaves.

**Reply by terjeio** [reply]
Ok, I support the STM32CubeIDE, some uses PlatformIO - likely with VSCode or even via the command line?

**Reply by olhodneland** [reply]
Ok. I'm sure I can use STM32CUBE or command line. Whatever is necessary. 

So what do I need from you to make it work on the uStepper S32 board and what are your suggested steps to get it working on the board?

**Reply by terjeio** [reply]
In the  STM32CubeIDE select _Debug F401 Blackpill_ from the "hammer" tool, then rigth click on the project and select _Properties_ from the popup-menu. Then select _C/C++ General > Paths and Symbols > Symbols_ and add `HSE_VALUE` with a value of 20000000 to GNU C:

![image](https://github.com/grblHAL/core/assets/20260062/8d3c8470-220f-4965-8fe1-3ceac88d4e5a)

You will also have to add a board map and modify _driver.h_ and _my_machine.h_ to use that.

**Reply by olhodneland** [reply]
In the driver.h files the comment in the beginning says: NOTE: do NOT change configurations here - edit my_machine.h instead!

So is that what I have to do?

**Reply by terjeio** [reply]
Adding a new board map to _driver.h_ is ok, changing other bits is not. You can also name your board map _my_machine_map.h_ to avoid changing _driver.h_.

**Reply by olhodneland** [reply]
The uStepper board has a TMC5130 stepper driver. I can only find TMC5160 in the git repo. Is it support for the TMC5130?

**Reply by terjeio** [reply]
> Is it support for the TMC5130?

The TMC5160 driver code can be used? The TMC5130 datasheet has to be compared with it or just try?

**Reply by olhodneland** [reply]
I looked at it and it seems like the difference is the voltage and current it can handle. The 5160 is much higher rated

**Reply by terjeio** [reply]
If the command set and structures are the same as the TMC5160 then the driver can be used.

**Reply by olhodneland** [reply]
I have done what you said but the board does not show up as a usb device after flashing it. I mean it does not show up outside the STM32Cube so I can connect to a gcode sender.

Any idea what is wrong?

**Reply by terjeio** [reply]
USB peripheral clock is wrong? If not 48MHz it will not work.

![image](https://github.com/grblHAL/core/assets/20260062/77058614-dc0e-4dd1-aa60-b6a9fceedec4)

The 401 clock setup in the driver should work if [HSE_VALUE is set to 20000000](https://github.com/grblHAL/STM32F4xx/blob/7d6dd894975347660ffe3735a961e3bf53ce8b4b/Src/main.c#L275) in the build configuration for _Debug F401 Blackpill_). The default seems to be 25000000 and that will not work. You may change the source code directly if you find that easier.

**Reply by olhodneland** [reply]
The HSE_VALUE is set to 20000000... still does not show up as a usb device after flashing. 
Could it be possible with a teams meeting or something to find out what is wrong, maybe I missed something🤷? I can share my screen and maybe we can find the problem?
I need this for a project quite soon.

**Reply by terjeio** [reply]
Teams is not working for me - I get a white screen I cannot get rid of, TeamViewer is ok. Do you have TeamViever installed?

**Reply by olhodneland** [reply]
Ok, i dont have teamviewer but i can download it. As long as i dont need paid version. I have anyDesk

**Reply by terjeio** [reply]
Sen me your email on SMS so I can use the web version of Teams.
